### PR TITLE
ARROW-564: [Python] Add Array.to_numpy()

### PR DESCRIPTION
**Pull request by kynan**
This PR came out of the [2018 AHL Hackathon](https://www.ahl.com/hackathon). I don't have much experience with the arrow code base.

### CODE REVIEW

**Review by pitrou — CHANGES_REQUESTED** [review]

File: python/pyarrow/array.pxi [hunk 1]
```diff
@@ -524,6 +528,20 @@ cdef class Array:
...
+        """
+        if self.null_count:
+            raise NotImplementedError('NumPy array conversion is only '
+                                      'supported for arrays without nulls.')
+        if not is_primitive(self.type.id):
+            raise NotImplementedError('NumPy array conversion is only '
+                                      'supported for primitive types.')
+        buflist = self.buffers()
+        return np.frombuffer(buflist[-1], dtype=self.type.to_pandas_dtype())[
+            self.offset:self.offset + len(self)]
```
Comment (wesm) [inline python/pyarrow/array.pxi]: This feels like a temporary workaround. If we accept this patch, can you open a JIRA about implementing this in libarrow_python (i.e. in C++) where we can utilize common code paths with `to_pandas` and toggle between NumPy-for-pandas and NumPy-for-NumPy behavior (and use the zero_copy_only flag where needed)?

File: python/pyarrow/tests/test_array.py [hunk 2]
```diff
@@ -83,6 +83,33 @@ def test_long_array_format():
...
+    import gc
+
+    arr = pa.array(range(10))
+
+    for i in range(10):
+        np_arr = arr.to_numpy()
+        assert sys.getrefcount(np_arr) == 2
+        np_arr = None  # noqa
+
+    assert sys.getrefcount(arr) == 2
```
[the remark this answers is not in the document]
Comment (kynan) [reply python/pyarrow/tests/test_array.py]: Done

File: python/pyarrow/tests/test_array.py [hunk 3]
```diff
@@ -83,6 +83,33 @@ def test_long_array_format():
...
+
+    for i in range(10):
+        arr = pa.array(range(10))
+        np_arr = arr.to_numpy()
+        arr = None
+        gc.collect()
+
+        # Ensure base is still valid
```
[the remark this answers is not in the document]
Comment (kynan) [reply python/pyarrow/tests/test_array.py]: Done

File: python/pyarrow/tests/test_array.py [hunk 4]
```diff
@@ -577,6 +604,31 @@ def test_simple_type_construction():
...
+        np.arange(10, dtype=np.uint8),
+        np.arange(10, dtype=np.float64),
+        np.arange(10, dtype=np.float32),
+        np.arange(10, dtype=np.float16),
+    ]
+)
+def test_to_numpy_roundtrip(narr):
+    arr = pa.array(narr)
+    assert narr.dtype == arr.to_numpy().dtype
+    assert np.array_equal(narr, arr.to_numpy())
```
[the remark this answers is not in the document]
Comment (kynan) [reply python/pyarrow/tests/test_array.py]: Done

File: python/pyarrow/tests/test_array.py [hunk 5]
```diff
@@ -83,6 +83,33 @@ def test_long_array_format():
     assert result == expected
 
 
+def test_to_numpy_zero_copy():
```
[the remark this answers is not in the document]
Comment (kynan) [reply python/pyarrow/tests/test_array.py]: This was based on `test_to_pandas_zero_copy` below. I've changed it according to your suggestions (which do make a lot more sense).

File: python/pyarrow/array.pxi [hunk 6]
```diff
@@ -599,6 +599,20 @@ cdef class Array:
                                               self, &out))
         return wrap_array_output(out)
 
+    def to_numpy(self):
+        """
+        Convert to a NumPy array
```
Comment (pitrou) [inline python/pyarrow/array.pxi]: The docstring should mention the zero-copy nature. For example "Construct a Numpy view of this array".

File: python/pyarrow/array.pxi [hunk 7]
```diff
@@ -599,6 +599,20 @@ cdef class Array:
...
+        """
+        if self.null_count:
+            raise NotImplementedError('NumPy array conversion is only '
+                                      'supported for arrays without nulls.')
+        if not is_primitive(self.type.id):
+            raise NotImplementedError('NumPy array conversion is only '
+                                      'supported for primitive types.')
+        buflist = self.buffers()
+        return np.frombuffer(buflist[-1], dtype=self.type.to_pandas_dtype())[
```
Comment (pitrou) [inline python/pyarrow/array.pxi]: Should we assert something about `len(buflist)` here?

File: python/pyarrow/tests/test_array.py [hunk 8]
```diff
@@ -87,6 +87,24 @@ def test_long_array_format():
...
+    np_arr[0] = 1
+    assert arr[0] == 1
+
+    assert sys.getrefcount(arr) == old_refcount
+
+    arr = None
+    import gc
+    gc.collect()
+
+    # Ensure base is still valid
```
Comment (pitrou) [inline python/pyarrow/tests/test_array.py]: Can you check the whole array contents here?

**Review by pitrou — CHANGES_REQUESTED** [review]
Thanks for the update! A few more comments below.

File: python/pyarrow/array.pxi (hunk 6 above)
Comment (kynan) [reply python/pyarrow/array.pxi]: Done

File: python/pyarrow/array.pxi (hunk 7 above)
Comment (kynan) [reply python/pyarrow/array.pxi]: Done

File: python/pyarrow/tests/test_array.py (hunk 8 above)
Comment (kynan) [reply python/pyarrow/tests/test_array.py]: Done

File: python/pyarrow/array.pxi (hunk 7 above)
Comment (pitrou) [reply python/pyarrow/array.pxi]: I was thinking `len(buflist) == 2` actually.

**Review by pitrou — APPROVED** [review]

**Comment by pitrou** [comment]
@kynan, thanks for your contribution! Can you open a JIRA account at https://issues.apache.org so that we can assign the issue to you?

File: python/pyarrow/array.pxi [hunk 9]
```diff
@@ -599,6 +599,21 @@ cdef class Array:
...
+        if self.null_count:
+            raise NotImplementedError('NumPy array view is only supported '
+                                      'for arrays without nulls.')
+        if not is_primitive(self.type.id):
+            raise NotImplementedError('NumPy array view is only supported '
+                                      'for primitive types.')
+        buflist = self.buffers()
+        assert len(buflist) == 2
+        return np.frombuffer(buflist[-1], dtype=self.type.to_pandas_dtype())[
+            self.offset:self.offset + len(self)]
```
Comment (wesm) [inline python/pyarrow/array.pxi]: This won't work on booleans, which are considered primitive by `is_primitive`. I'll write a quick patch to test/fix

**Comment by wesm** [comment]
I opened https://issues.apache.org/jira/browse/ARROW-2869 about adding some Sphinx documentation for this

**Comment by wesm** [comment]
Unfortunately, this patch does not satisfy the scope of ARROW-564. I will open a new JIRA to assert the desired API re: nulls

**Comment by wesm** [comment]
See https://issues.apache.org/jira/browse/ARROW-2870

**Comment by kynan** [comment]
@pitrou I have an account on Apache JIRA (different user name though, seems you found it).

**Pull request closed without merging**